src/concrete/ml/onnx/onnx_utils.py

```diff
@@ -495,3 +507,91 @@ def remove_initializer_from_input(model: onnx.ModelProto):  # pragma: no cover
             inputs.remove(name_to_input[initializer.name])
 
     return model
+
+
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml/issues/397
+def compute_lsb_to_remove_for_trees(onnx_model: onnx.ModelProto, q_x: numpy.ndarray) -> List[int]:
+    """Compute the LSB to remove for trees in the first and second stage.
+
+    Args:
+        onnx_model (onnx.ModelProto): The model to clean
+        q_x (numpy.ndarray): The quantized inputs
+
+    Returns:
+        List: the number of LSB to remove for stage 1 and stage 2
+    """
+
+    def get_bitwidth(array: numpy.ndarray) -> int:
+        """Compute the bitwidth required to represent the largest value in `array`.
+
+        Args:
+            array (umpy.ndarray): The array for which the bitwidth needs to be checked.
+
+        Returns:
+            int: The required bits to represent the array.
+        """
+
+        max_val = numpy.max(numpy.abs(array))
+        bitwidth = math.ceil(math.log2(max_val + 1)) + 1
+        return bitwidth
+
+    def update_lsbs_if_overflow_detected(array: numpy.ndarray, initial_bitwidth: int) -> int:
+        """Update the number of LSBs to remove based on overflow detection.
+
+        Args:
+            array (umpy.ndarray): The array for which the bitwidth needs to be checked.
+            initial_bitwidth (int): The target bitwidth that should not be exceeded.
+
+        Returns:
+            int: The updated LSB to remove.
+        """
+
+        lsbs_to_remove = initial_bitwidth
+
+        if lsbs_to_remove > 0:
+            half = 1 << (lsbs_to_remove - 1)
+            if get_bitwidth(array - half) <= initial_bitwidth:
+                lsbs_to_remove -= 1
+
+        return lsbs_to_remove
+
+    quant_params = {
```

Comment:
from here down to the end of the function : maybe add some comments to explain a bit more what is going on !